BootloaderCommonPkg/Library/ContainerLib/ContainerLib.c

```diff
@@ -0,0 +1,523 @@
+{
+  EFI_STATUS  Status;
+
+  if (AuthType == AUTH_TYPE_SHA2_256) {
+    Status = DoHashVerify (Data, Length, HASH_TYPE_SHA256, CompType, HashData);
+  } else if (AuthType == AUTH_TYPE_SIG_RSA2048_SHA256) {
+    Status = DoRsaVerify (Data, Length, CompType, AuthData,
+                          AuthData + RSA2048NUMBYTES, HashData, NULL);
+  } else if (AuthType == AUTH_TYPE_NONE) {
+    Status = EFI_SUCCESS;
+  } else {
+    Status = EFI_UNSUPPORTED;
+  }
+  return Status;
+}
+
+/**
+  Locate a component from a container.
+
+  @param[in]     ContainerSig       Container signature.
+  @param[in]     ComponentName      Component name.
+  @param[in,out] ContainerEntryPtr  Pointer to receive container entry info.
+  @param[in,out] ComponentEntryPtr  Pointer to receive component entry info.
+
+  @retval EFI_UNSUPPORTED          Unsupported AuthType.
+  @retval EFI_NOT_FOUND            Cannot locate component.
+  @retval EFI_SECURITY_VIOLATION   Authentication failed.
+  @retval EFI_SUCCESS              component region is located successfully.
+
+**/
+EFI_STATUS
+LocateComponent (
+  IN      UINT32                  ContainerSig,
+  IN      UINT32                  ComponentName,
+  IN OUT  CONTAINER_ENTRY       **ContainerEntryPtr,
+  IN OUT  COMPONENT_ENTRY       **ComponentEntryPtr
+  )
+{
+  EFI_STATUS                Status;
+  CONTAINER_HDR            *ContainerHdr;
+  CONTAINER_ENTRY          *ContainerEntry;
+  COMPONENT_ENTRY          *CompEntry;
+  UINT8                    *AuthData;
+  UINT32                    ContainerBase;
+  UINT32                    ContainerSize;
+  UINT32                    ContainerHdrSize;
+  UINT8                     AuthType;
+
+  // Search container header from cache
+  ContainerEntry = GetContainerBySignature (ContainerSig);
+  if (ContainerEntry == NULL) {
+    // Location container from Flash Map
+    Status = GetComponentInfo (ContainerSig, &ContainerBase, &ContainerSize);
+    if (EFI_ERROR (Status)) {
+      return EFI_NOT_FOUND;
+    }
+
+    // Register container temporarily
+    DEBUG ((EFI_D_INFO, "Register container 0x%08X\n", ContainerSig));
+    Status = RegisterContainer (ContainerBase);
+    if (EFI_ERROR (Status)) {
+      return EFI_UNSUPPORTED;
+    }
+
+    // Find authentication data offset and authenticate the container header
+    ContainerEntry   = GetContainerBySignature (ContainerSig);
+    ContainerHdr     = (CONTAINER_HDR *)ContainerEntry->HeaderCache;
+    ContainerHdrSize = GetContainerHeaderSize (ContainerHdr);
+    if (ContainerHdrSize > 0) {
+      AuthType = ContainerHdr->AuthType;
+      AuthData = (UINT8 *)ContainerHdr + ALIGN_UP(ContainerHdrSize, AUTH_DATA_ALIGN);
+      if ((AuthType == AUTH_TYPE_NONE) && FeaturePcdGet (PcdVerifiedBootEnabled)) {
+        Status = EFI_SECURITY_VIOLATION;
+      } else {
+        Status = AuthenticateComponent ((UINT8 *)ContainerHdr, ContainerHdrSize,
+                                             AuthType, AuthData, NULL, COMP_TYPE_PUBKEY_FWU);
```

Comment:
To verify the container header, the authenticate method should not depend on container itsself since it is not been verified yet.  In this case, if container st the AuthType to None, container could bypass verification.
I think we could verify the container header by RSA2048 if hashstore is using SHA256 or by RSA3072 if hashstore is using SHA384.

Reply:
We do verify the AuthType separately.  When it is NONE,  it will pass the container header check only if VerifiedBoot is disabled.

Reply:
OK. I checked the DoRsaVerify() and DoHashVerify().  To verify the public key of container, it is still using hardcode sha256 now, not using AuthType.